# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** 10/08/2025

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [3]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse
library(tidyverse)

# Set working directory


# Import datasets with exact variable names
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
head(customers, 3)
head(orders, 3)
head(order_items, 3)
head(products, 3)
head(suppliers, 3)

Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): OrderID, CustomerID, Total_Amount
date (1): Order_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 400 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (4): OrderID, ProductID, Quantity, Unit_Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this m

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Order Items: 400 rows x 4 columns
Products: 50 rows x 4 columns
Suppliers: 10 rows x 3 columns


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20

OrderID CustomerID Order_Date Total_Amount
1 1       87         2023-08-30 424.30      
2 2       12         2024-03-24 183.09      
3 3       37         2024-03-19 549.07

OrderID ProductID Quantity Unit_Price
1 213      8        1         43.81    
2 176     18        5        489.16    
3 118      2        5        442.09

ProductID Product_Name Category    Supplier_ID
1 1         Product 1    Home        6          
2 2         Product 2    Electronics 7          
3 3         Product 3    Home        1

Supplier_ID Supplier_Name Country
1 1           Supplier 1    USA    
2 2           Supplier 2    USA    
3 3           Supplier 3    Japan

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [ ]:
# Your code here:
# Create an inner join between customers and orders
# Analyze the results and compare row counts



In [ ]:
# Your code here:
# Create a left join to keep all customers
# Count how many customers have no orders



In [ ]:
# Your code here:
# Create a right join to keep all orders
# Check for orders with invalid customer IDs



In [ ]:
<VSCode.Cell id="#VSC-0da36c1e" language="markdown">
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

**Part 2.1:** Create `customer_orders` using inner_join()
**Part 2.2:** Create `customer_orders_left` using left_join()  
**Part 2.3:** Create `customer_orders_right` using right_join()
**Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

In [4]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- inner_join(customers, orders, by = "CustomerID")
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [6]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:\n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [ ]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:


# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

In [ ]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:


# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

In [7]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:\n")
cat("Total rows:", nrow(customer_orders_right), "\n")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [8]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
customer_orders_full <- full_join(customers, orders, by = "CustomerID")
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:\n")
cat("Total rows:", nrow(customer_orders_full), "\n")
cat("Customers without orders:", customers_only, "\n")
cat("Orders without valid customers:", orders_only, "\n")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [ ]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:


# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [9]:
# Step 1: REQUIRED variable name: orders_items
orders_items <- inner_join(orders, order_items, by = "OrderID")
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [10]:
# Step 2: REQUIRED variable name: orders_customers_items  
orders_customers_items <- inner_join(orders_items, customers, by = "CustomerID")
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [11]:
# Step 3: REQUIRED variable name: complete_order_data
complete_order_data <- inner_join(orders_customers_items, products, by = "ProductID")
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 310 rows


In [13]:
# Step 4: REQUIRED variable name: complete_data
complete_data <- inner_join(complete_order_data, suppliers, by = "Supplier_ID")
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [ ]:
<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

In [14]:
# REQUIRED variable name: customers_no_orders
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [15]:
# REQUIRED variable name: orphaned_orders
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [16]:
# REQUIRED variable name: products_never_ordered
products_never_ordered <- anti_join(products, order_items, by = "ProductID")
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [17]:
# REQUIRED variable name: active_customers
active_customers <- semi_join(customers, orders, by = "CustomerID")
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [19]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset

duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [ ]:
<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

In [22]:
# REQUIRED variable name: customer_metrics
customer_metrics <- orders %>%
  left_join(order_items, by = "OrderID") %>%
  left_join(customers, by = "CustomerID") %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Quantity * Unit_Price, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    Avg_Order_Value = ifelse(Order_Count > 0, Total_Spent / Order_Count, NA_real_),
    .groups = "drop"
  )


cat("Customer Analysis Summary:\n")
cat("Total customers analyzed:", nrow(customer_metrics), "\n")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "\n")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary:


Total customers analyzed: 105 
Top customer total spent: $ 17221.69 
# A tibble: 3 × 5
  CustomerID Name  Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>       <dbl>       <int>           <dbl>
1        103 NA         17222.          10           1722.
2        102 NA         16116.          10           1612.
3        104 NA         13427.          10           1343.


In [23]:
# REQUIRED variable name: product_metrics  
product_metrics <- order_items %>%
  left_join(products, by = "ProductID") %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  )


cat("Product Analysis Summary:\n")
cat("Total products analyzed:", nrow(product_metrics), "\n")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "\n")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 13970.66 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        26 Product 26          13971.             60              18
2        47 Product 47          12582.             38              10
3        43 Product 43          10584.             36              10


In [24]:
# REQUIRED variable name: supplier_metrics
supplier_metrics <- products %>%
  left_join(suppliers, by = "Supplier_ID") %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID),
    .groups = "drop"
  )


cat("Supplier Analysis Summary:\n")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "\n")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "\n")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [25]:
# REQUIRED variable name: regional_analysis
regional_analysis <- orders %>%
  left_join(order_items, by = "OrderID") %>%
  left_join(customers, by = "CustomerID") %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, NA_real_),
    .groups = "drop"
  )


cat("Regional Analysis Summary:\n")
cat("Total cities analyzed:", nrow(regional_analysis), "\n")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "\n")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 6 
Highest city sales: $ 72277.52 
# A tibble: 6 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             24              3012.
2 NA               68753.              5             13751.
3 Phoenix          49333.             19              2596.
4 Los Angeles      48506.             22              2205.
5 Houston          38143.             19              2008.
6 New York         36417.             16              2276.


In [ ]:
# Your code here:
# Analyze product performance metrics



In [ ]:
# Your code here:
# Evaluate supplier performance



In [ ]:
# Your code here:
# Create regional analysis by customer city



In [26]:
# Supplier Performance Analysis
supplier_metrics <- complete_supply_chain %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

ERROR: Error: object 'complete_supply_chain' not found


In [27]:
# Regional Analysis by Customer City
regional_analysis <- complete_supply_chain %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

ERROR: Error: object 'complete_supply_chain' not found


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [28]:
# REQUIRED variable name: top_customers
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
top_customers <- customer_metrics %>%
  filter(Total_Spent >= top_10_percent_threshold)

cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 5546.35 
Number of top customers: 11 


In [30]:
# REQUIRED variable name: product_combinations
product_combinations <- order_items %>%
  group_by(OrderID) %>%
  filter(n() > 1) %>%
  arrange(OrderID) %>%
  left_join(products, by = "ProductID") %>%
  group_by(OrderID) %>%
  summarise(Combo = paste(sort(unique(Product_Name)), collapse = " & "), .groups = "drop") %>%
  count(Combo, name = "n") %>%
  arrange(desc(n))
  cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
  cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 119 
Most frequent combination appears 2 times


In [31]:
# REQUIRED variable name: critical_suppliers
critical_suppliers <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Product_Count = n_distinct(ProductID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue), desc(Product_Count))

cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))


Critical Suppliers Analysis:
Total suppliers analyzed: 10 
[1] "Top 3 most critical suppliers:"
# A tibble: 3 × 4
  Supplier_ID Supplier_Name Total_Revenue Product_Count
        <dbl> <chr>                 <dbl>         <int>
1           5 Supplier 5           43434.             5
2           7 Supplier 7           35416.            10
3          10 Supplier 10          31875.             6


In [32]:
# REQUIRED variable name: market_expansion  
market_expansion <- regional_analysis %>%
  filter(Customer_Count > 1) %>%
  arrange(Avg_Customer_Value)

cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 6 
[1] "Top expansion opportunities:"
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Houston          38143.             19              2008.
2 Los Angeles      48506.             22              2205.
3 New York         36417.             16              2276.
4 Phoenix          49333.             19              2596.
5 Chicago          72278.             24              3012.


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** The most efficient join seemed to be inner join due to having multiplematching records on different tables. 
- **Data Relationships:** Data relationships were consistent, just like in excel and Access. Primary key in one table matched foreign key in another. 
- **Multi-table Joins:** The multi-table joins provided a comprehensive view of the data, revealing insights that were not apparent in single-table analyses.

### Data Quality Issues Discovered:
- **Orphaned Records:** There seemed to be 50 orphaned records in the orders table that did not have matching customer IDs.
- **Missing Data:** Regional analysis and customer analysis were impacted by missing city information in 10% of customer records.
- **Referential Integrity:** Orders table and order items table both had foreign key violations with 5 records in order items not matching any order ID. Mostly when an anti join command is used between a child table and a parent table as well. 

### Business Insights:
- **Top Customer:** The top customer seemed to be customer ID 103 with 17222 in total spending. 
- **Best Product:** Product 26 with 13971 in revenue, generated the most revenue.
- **Regional Performance:** [Key findings about city-level sales patterns]
- **Supplier Analysis:** [Insights about supplier diversity and concentration]

### Strategic Recommendations:
- 3 strategic recomendations would include: targeting the top percentage of customers with loyalty programs, focusing marketing efforts on high-performing products, and exploring expansion opportunities in underperforming cities with high customer counts.

### Technical Learnings:
- **Inner vs Left Joins:** Inner joins were used to analyze relationships where only matching records were needed, such as finding customers with orders or products that were actually sold. Left joins were used when it was important to retain all records from the primary table, such as listing all customers (including those without orders) or all products (including those never ordered).
- **Multi-table Strategy:** Multi-table joins were built stepwise, starting with two-table joins and progressively adding more tables (orders + order_items → + customers → + products → + suppliers). This approach ensured data integrity at each stage and allowed for targeted analysis at different points in the supply chain.
- **Performance Considerations:** Inner joins were more efficient for large datasets since they reduced the result size by excluding unmatched records. Left and full joins increased data size and required more memory, especially when including all records from both tables. Careful use of anti_join and semi_join helped identify data quality issues without unnecessary data expansion.

In [ ]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")